src/main/java/org/truffleruby/core/thread/ThreadManager.java

```diff
@@ -50,35 +50,52 @@
 
 public class ThreadManager {
 
+    public static final String NAME_PREFIX = "Ruby Thread";
+
     private final RubyContext context;
 
     private final DynamicObject rootThread;
-    private final ThreadLocal<DynamicObject> currentThread = new ThreadLocal<>();
+    private final Thread rootJavaThread;
+
+    private final Map<Thread, DynamicObject> foreignThreadMap = new ConcurrentHashMap<>();
+    private final ThreadLocal<DynamicObject> currentThread = ThreadLocal.withInitial(() -> foreignThreadMap.get(Thread.currentThread()));
+
+    private final Set<DynamicObject> runningRubyThreads =
+            Collections.newSetFromMap(new ConcurrentHashMap<DynamicObject, Boolean>());
 
-    private final Set<DynamicObject> runningRubyThreads
-            = Collections.newSetFromMap(new ConcurrentHashMap<DynamicObject, Boolean>());
+    private final Set<Thread> threadsWeCreated =
```

Comment:
Is there a more descriptive name for this? E.g.  `rubyManagedThreads`, `runningRubyManagedThreads`

Reply:
:+1: `rubyManagedThreads` sounds good and is consistent with the predicate method.

Reply:
Done.